File: next_permutation.py
from typing import List
class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        length = len(nums)
        if length ==1:
            return
        
        i = length-2
        while(i>=0):
            if nums[i] < nums[i+1]:
                break
            i-=1

        if i<0:
            nums.reverse()
            return
        
        j = length-1
        while j > i:
            if nums[j] > nums[i]:
                break
            j-=1
        
        nums[i], nums[j] = nums[j], nums[i]

        nums[i+1:] = nums[i+1:][::-1]
k = [2,3,1]
sol = Solution().nextPermutation(k)
print(k)

File: next_permutation.go
package main

import "fmt"

func main() {
	arr := []int{1, 5, 1}
	nextPermutation(arr)
	fmt.Print(arr)
}

func nextPermutation(nums []int) {
	lastIndex := len(nums) - 1

	if lastIndex == 0 {
		return
	}
	i := lastIndex - 1
	for ; i >= 0; i-- {
		if nums[i] < nums[i+1] {
			break
		}
	}

	if i < 0 {
		reverse(nums, 0, lastIndex)
		return
	}
	j := lastIndex
	for ; j > i && nums[j] <= nums[i]; j-- {
	}

	nums[i], nums[j] = nums[j], nums[i]

	reverse(nums, i+1, lastIndex)

}

func reverse(nums []int, left, right int) {
	for left < right {
		nums[left], nums[right] = nums[right], nums[left]
		left++
		right--
	}
}
